tests/e2e/tests/test_kiali_service_endpoint.py

```diff
@@ -1,46 +1,138 @@
 import pytest
 import tests.conftest as conftest
+from utils.command_exec import command_exec
 
 BOOKINFO_EXPECTED_SERVICES = 4
 BOOKINFO_EXPECTED_SERVICES_MONGODB = 5
 SERVICE_TO_VALIDATE = 'reviews'
+VIRTUAL_SERVICE_FILE = 'assets/bookinfo-reviews-80-20.yaml'
+DESTINATION_RULE_FILE = 'assets/bookinfo-destination-rule-reviews.yaml'
 
 def test_service_list_endpoint(kiali_client):
     bookinfo_namespace = conftest.get_bookinfo_endpoint()
 
-    assert kiali_client.service_list(namespace=bookinfo_namespace).get('namespace').get('name') == bookinfo_namespace
-    number_services = len (kiali_client.service_list(namespace=bookinfo_namespace).get('services'))
-    assert (number_services == BOOKINFO_EXPECTED_SERVICES) or (number_services == BOOKINFO_EXPECTED_SERVICES_MONGODB)
+    service_list_json = kiali_client.service_list(namespace=bookinfo_namespace)
+    assert service_list_json.get('namespace').get('name') == bookinfo_namespace
+
+    services = service_list_json.get('services')
+    assert (len (services) == BOOKINFO_EXPECTED_SERVICES) or (len (services) == BOOKINFO_EXPECTED_SERVICES_MONGODB)
+
+    for service in services:
+      assert service != None
+      assert service.get('name') != None and service.get('name') != ''
+      assert service.get('istioSidecar') == True
+      assert service.get('appLabel') == True
 
 def test_service_detail_endpoint(kiali_client):
     bookinfo_namespace = conftest.get_bookinfo_endpoint()
 
     service_details = kiali_client.service_details(namespace=bookinfo_namespace, service=SERVICE_TO_VALIDATE)
     assert service_details != None
     assert service_details.get('istioSidecar') == True
+    assert 'ports' in service_details.get('service')
+    assert 'labels' in service_details.get('service')
     assert 'workloads' in service_details
     assert 'service' in service_details
+    assert 'endpoints' in service_details
+    assert 'virtualServices' in service_details
+    assert 'destinationRules' in service_details
+    assert 'health' in service_details
+
+def test_service_detail_with_virtual_service(kiali_client):
+    bookinfo_namespace = conftest.get_bookinfo_endpoint()
+
+    # Add a virtual service that will be tested
+    assert command_exec.oc_apply(VIRTUAL_SERVICE_FILE, bookinfo_namespace) == True
+
```

Comment:
If the deployed VS has not yet propagated through Isio yet, without a check that the VS is actually present, could there be false failures below this point due to timing issues?

Reply:
you are right. Do you suggest to add the assertions within a timeout? Similarly to what you have in  graph_badges test?

Reply:
The graph_badge tests can be an example. A type of polling loop that checks that the operation was completed, using a timeout-class. An assertion is not necessary, just a check for a successfully completion. The timeout class will throw an exception if the polling for the condition does not break.